### PR TITLE
Feature session persistence redis

### PR DESCRIPTION
**Pull request by sthenusan**
### Proposed changes in this pull request

Currently, the identity server user session implementation is backed by relational databases. That implementation will affect the performance in large scale implementations. This pull request contains POC to the session data is backed by Redis server. Here Both Redis and RDMS support is given.

#### Code
The persistSessionData,  getSessionContextData methods in the Session store class are modified for the change. enableRedis variable is used to select either the Redis implementation or RDMS implementation.

### CODE REVIEW

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/SessionDataStore.java [hunk 1]
```diff
@@ -145,6 +150,8 @@
     private boolean sessionDataCleanupEnabled = true;
     private boolean operationDataCleanupEnabled = false;
     private static boolean tempDataCleanupEnabled = false;
+    private static BinaryJedis jedis;
+    private boolean enableRedis= false;
```
Comment (ChamathNS) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/SessionDataStore.java:154]: When are we changing this boolean value?

Comment (ChamathNS) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/SessionDataStore.java:154]: Shall we use a more comprehensible variable name? The current name doesn't clearly denote if it is a boolean IMO.

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/SessionDataStore.java [hunk 2]
```diff
@@ -486,15 +455,89 @@ public void removeExpiredOperationData() {
...
         if (!enablePersist) {
-            return;
+            return null;
         }
-        Connection connection = null;
-        try {
-            connection = IdentityDatabaseUtil.getDBConnection();
-        } catch (IdentityRuntimeException e) {
-            log.error(e.getMessage(), e);
+        if (enableRedis){
```
Comment (ChamathNS) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/SessionDataStore.java:470]: Not formatted.

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/SessionDataStore.java [hunk 3]
```diff
@@ -486,15 +455,89 @@ public void removeExpiredOperationData() {
...
+                    }
+                }
+            } catch (ClassNotFoundException | IOException | SQLException |
+                    IdentityApplicationManagementException e) {
+                log.error("Error while retrieving session data", e);
+            } finally {
+                IdentityDatabaseUtil.closeAllConnections(connection, resultSet, preparedStatement);
+            }
+            return null;
+
```
Comment (ChamathNS) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/SessionDataStore.java:533]: Excess new line.

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/SessionDataStore.java [hunk 4]
```diff
@@ -288,61 +295,23 @@ public static SessionDataStore getInstance() {
         return instance;
     }
 
+    //Return Jedis object
+    private static BinaryJedis getJedisInstance() {
+
+        if (jedis != null)
+            return jedis;
+        return new BinaryJedis("127.0.0.1", 6379);
```
[the remark this answers is not in the document]
Comment (sthenusan) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/store/SessionDataStore.java]: corrected

**Comment by sthenusan** [comment]
https://github.com/wso2/carbon-identity-framework/pull/3279

**Comment by sthenusan** [comment]
new PR https://github.com/wso2/carbon-identity-framework/pull/3279

**Pull request closed without merging**